### PR TITLE
Cypher filtering many to many relationships

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR adds the ability to filter on many to many (m:m) relationship cypher fields. An example:

Type definitions:
```gql
type Movie @node {
    title: String
    actors: [Actor!]!
        @cypher(
            statement: """
            MATCH (this)<-[:ACTED_IN]-(actor:Actor)
            RETURN actor
            """
            columnName: "actor"
        )
}

type Actor @node {
    name: String
    movies: [Movie!]!
        @cypher(
            statement: """
            MATCH (this)-[:ACTED_IN]->(movie:Movie)
            RETURN movie
            """
            columnName: "movie"
        )
}
```

You can now query with a filter based on the `actors` field:
```gql
query {
  movies(where: { actors: { name: "Jada Pinkett Smith" } }) {
      title
  }
}
```

## Complexity

Medium

# Issue

Relates to #554, but not fully closing as this is LTS.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 36 | 48 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 45 | 69 | 1134 |
|  aggregations.NestedAggregation | 15407 | 15407 | 58 | 72 | 2174 |
|  aggregations.AggregationWithWhere | 10833 | 10833 | 47 | 66 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20097917 | 20097917 | 1906 | 2051 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20097917 | 20097917 | 1859 | 1973 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 8603 | 8603 | 31 | 36 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8791 | 8791 | 40 | 39 | 2174 |
|  aggregations.DeeplyNestedCount | 10052335 | 10052335 | 2459 | 2672 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 43 | 82 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 71 | 121 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 129 | 158 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 64 | 64 | 11 |
|  connect.createAndConnect | 6433 | 6433 | 144 | 181 | 3003 |
|  connections.Connection | 12951 | 12951 | 61 | 65 | 2174 |
|  connections.NestedConnection | 37705 | 37705 | 124 | 153 | 4516 |
|  create.SimpleMutation | 7 | 7 | 55 | 59 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 23 | 34 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 617 | 638 | 1040 |
|  delete.NestedDeleteInUpdate | 16844 | 16844 | 158 | 149 | 2040 |
|  2871.NestedRelationshipFilter | 19868 | 19735 | 49 | 75 | 4470 |
|  2925.SingleRelationshipFilter | 5245 | 5245 | 35 | 43 | 1040 |
|  2925.NestedSingleRelationshipFilter | 17641 | 17641 | 68 | 106 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 37 | 43 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 56 | 61 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 19 | 26 | 1040 |
|  query.SimpleQueryWithRelationship | 15031 | 15031 | 37 | 48 | 2174 |
|  query.QueryWhere | 8564 | 8564 | 33 | 42 | 2154 |
|  query.SimpleQueryWithNestedWhere | 8713 | 8713 | 45 | 55 | 2154 |
|  query.Nested | 10084891 | 10084891 | 6877 | 6852 | 2008534 |
|  query.NestedWithFilter | 10064992 | 10064992 | 6426 | 6562 | 2004000 |
|  query.OrFilterOnRelationships | 37354 | 37241 | 148 | 169 | 2169 |
|  query.OrFilterOnRelationshipsAndNested | 27427 | 27672 | 218 | 228 | 2169 |
|  query.QueryWithNestedIn | 14858 | 14948 | 52 | 71 | 2086 |
|  query.NestedConnectionWhere | 8703 | 8703 | 53 | 67 | 2174 |
|  query.DeeplyNestedConnectionWhere | 8883 | 8793 | 69 | 83 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 17693 | 17489 | 137 | 147 | 1664 |
|  query.NestedWithRelationshipSingleFilters | 3808 | 3808 | 141 | 195 | 1134 |
|  query.Fulltext | 64 | 64 | 31 | 36 | 16 |
|  query.FulltextWithNestedQuery | 516 | 516 | 39 | 44 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 12961 | 12961 | 41 | 44 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 12961 | 12961 | 62 | 68 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 13096 | 13096 | 56 | 58 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 13096 | 13096 | 93 | 100 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 13696 | 13658 | 98 | 126 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 13266 | 13266 | 97 | 131 | 2174 |
|  sorting.SortMultipleTypes | 3436 | 3436 | 94 | 112 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 13321 | 13321 | 96 | 128 | 2174 |
|  sorting.SortOnNestedFields | 12951 | 12951 | 40 | 51 | 2174 |
|  sorting.SortDeeplyNestedFields | 39785 | 39785 | 75 | 85 | 4516 |
|  sorting.ConnectionWithSort | 3271 | 3271 | 90 | 75 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 90 | 59 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 52 | 64 | 35 |
|  unions.NestedUnion | 309975 | 309975 | 256 | 270 | 33033 |
|  unions.NestedUnionWithMissingFields | 283949 | 283949 | 239 | 267 | 33033 |
|  update.NestedUpdate | 14137 | 14137 | 94 | 98 | 2002 |

</details>
        
Old Schema Generation: 27.557s
Schema Generation: 27.758s
Old Subgraph Schema Generation: 28.241s
Subgraph Schema Generation: 28.533s

File: packages/graphql/src/schema-model/generate-model.ts [hunk 1]
```diff
@@ -136,14 +136,29 @@ function hydrateCypherAnnotations(schema: Neo4jGraphQLSchemaModel, concreteEntit
         for (const attributeField of concreteEntity.attributes.values()) {
             if (attributeField.annotations.cypher) {
                 if (attributeField.type instanceof ObjectType) {
-                    const foundConcreteEntity = schema.getConcreteEntity(attributeField.type.name);
-                    if (!foundConcreteEntity) {
+                    const targetEntity = schema.getConcreteEntity(attributeField.type.name);
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/generate-model.ts]: I'm noticing this logic is already handled on the 7.x branch, I believe that you can avoid conflicts by using that!

File: packages/graphql/src/translate/queryAST/factory/FilterFactory.ts [hunk 2]
```diff
@@ -334,28 +335,33 @@ export class FilterFactory {
...
+
+            const filter = attribute.typeHelper.isList()
+                ? this.createCypherRelationshipFilterTreeNode(options)
+                : this.createCypherOneToOneRelationshipFilterTreeNode(options);
 
             if (!isNull) {
                 const entityWhere = where[concreteEntity.name] ?? where;
                 const targetNodeFilters = this.createNodeFilters(concreteEntity, entityWhere);
-                cypherOneToOneRelationshipFilter.addTargetNodeFilter(...targetNodeFilters);
+                filter.addTargetNodeFilter(...targetNodeFilters);
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/factory/FilterFactory.ts:358]: would not be better to accept these filters in the `CypherRelationshipFilter` constructor and invoke this.createNodeFilters from `this.createCypherRelationshipFilterTreeNode` or `this.createCypherOneToOneRelationshipFilterTreeNode`?

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+import { TestHelper } from "../../../../../utils/tests-helper";
+
+describe("Connection API - cypher directive filtering - relationship auth filter", () => {
+    const testHelper = new TestHelper();
+
+    afterEach(async () => {
+        await testHelper.close();
+    });
+
+    test("(relationship with auth filter on type PASS", async () => {
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]:
```suggestion
    test("relationship with auth filter on type PASS", async () => {
```

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        const Movie = testHelper.createUniqueType("Movie");
+        const Actor = testHelper.createUniqueType("Actor");
+
+        const typeDefs = /* GraphQL */ `
+            type ${Movie} @node @authorization(filter: [{ where: { node: { actors: { name: "$jwt.custom_value" } } } }]) {
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]:
```suggestion
            type ${Movie} @node @authorization(filter: [{ where: { node: { actors_SOME: { name: "$jwt.custom_value" } } } }]) {
```

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+            CREATE (a3)-[:ACTED_IN]->(m2)
+            CREATE (a3)-[:ACTED_IN]->(m3)
+            `,
+            {}
+        );
+
+        const query = /* GraphQL */ `
+            query {
+                ${Movie.operations.connection}(
+                    where: {
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]: I may be wrong, but I believe this is assumption is valid without or without any authorization, I would use a dataset that is filtered by the authorization rules

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+            CREATE (a3:${Actor} { name: "Jada Pinkett Smith" })
+            CREATE (a3)-[:ACTED_IN]->(m2)
+            CREATE (a3)-[:ACTED_IN]->(m3)
+            `,
+            {}
+        );
+
+        const query = /* GraphQL */ `
+            query {
+                ${Movie.operations.connection}(
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]: As the comment above, I think that for filter you can have a single test that has a dataset that is filtered by an authorization rule, in this case if you remove the `where: { title: "The Matrix" }` and pass `Keanu Revees` in the JWT then `The Matrix` should be returned. The test name seems implying that one test fail and the other pass, while in this case both are filtered!

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+                    where: {
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]: Also here, you can change the test by remove this `where` and put `Keanu Revees` in the JWT, this test should fail anyway!

File: packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TestHelper } from "../../../../../utils/tests-helper";
+
+describe("Connection API - cypher directive filtering - relationship auth filter", () => {
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/relationships/cypher-filtering-relationship-auth-connection.int.test.ts]: I think you can remove the whole suite, and apply the suggestions for the other suite `describe("cypher directive filtering - relationship auth filter`, I believe what is worth testing is the `Connection` filtering style such as:`@authorization(filter: [{ where: { node: { actorsConnection_SOME: { name: "$jwt.custom_value" } }` that is not something supported! Another valid case to test using Connection Projection is for `field-leve auth` but that again is something not supported!

**Review by MacondoExpress — COMMENTED** [review]
LGTM, left some comments mainly around tests!

File: packages/graphql/src/schema-model/generate-model.ts (hunk 1 above)
Comment (mjfwebb) [reply packages/graphql/src/schema-model/generate-model.ts]: I'm not sure what you mean, can you elaborate please?

Comment (MacondoExpress) [reply packages/graphql/src/schema-model/generate-model.ts]: Sure, I meant that on the 7.x branch this section of the code is different and I believe I already handled this case!

File: packages/graphql/src/translate/queryAST/factory/FilterFactory.ts (hunk 2 above)
Comment (mjfwebb) [reply packages/graphql/src/translate/queryAST/factory/FilterFactory.ts:358]: Will do this in the future when the relationshipFilter doesn't also do this.

**Review by MacondoExpress — APPROVED** [review]

**Pull request merged**